metagraph/src/graph/representation/succinct/boss_chunk_construct.cpp

```diff
@@ -673,114 +710,225 @@ void recover_dummy_nodes_disk(KmerCollector &kmer_collector,
-            dummy_next_names[i] = dir/fmt::format("dummy_source_{}_{}", dummy_pref_len + 1, i);
-            dummy_next_chunks.emplace_back(dummy_next_names[i], ENCODER_BUFFER_SIZE);
-        }
-
-        KMER prev_kmer(0);
+        // this will store blocks of source dummy k-mers of the current level
         uint64_t num_kmers = 0;
-        const std::function<void(const KMER_INT &)> &write_dummy = [&](const KMER_INT &v) {
-            KMER kmer(v);
-            kmer.to_prev(k + 1, BOSS::kSentinelCode);
-            if (prev_kmer != kmer) {
-                dummy_next_chunks[kmer[0]].add(kmer.data());
-                prev_kmer = std::move(kmer);
+        result.emplace_back(alphabet_size);
+        std::vector<std::string> next_names(alphabet_size * alphabet_size);
+        #pragma omp parallel for num_threads(num_threads) schedule(dynamic)
+        for (TAlphabet F = 0; F < alphabet_size; ++F) {
+            // will store sorted dummy source k-mers of pattern ***F*
+            result.back()[F] = dir/fmt::format("dummy_source_{}_{}", dummy_pref_len, F);
+            Encoder<KMER_INT> dummy_chunk(result.back()[F], ENCODER_BUFFER_SIZE);
+
+            std::vector<Encoder<KMER_INT>> dummy_next_chunks;
+            for (TAlphabet i = 0; i < alphabet_size; ++i) {
+                next_names[i * alphabet_size + F]
+                    = dir/fmt::format("dummy_source_{}_{}_{}", dummy_pref_len + 1, i, F);
+                dummy_next_chunks.emplace_back(next_names[i * alphabet_size + F],
+                                               ENCODER_BUFFER_SIZE);
             }
-            num_kmers++;
-        };
-        common::merge_files(dummy_names, write_dummy, false);
 
-        std::for_each(dummy_next_chunks.begin(), dummy_next_chunks.end(),
-                      [](auto &v) { v.finish(); });
-        dummy_names = std::move(dummy_next_names);
+            KMER prev_kmer(0);
+            const std::function<void(const KMER_INT &)> &write_dummy = [&](const KMER_INT &v) {
+                dummy_chunk.add(v);
+                // ***F*
+                KMER kmer(v);
+                // $***F
+                kmer.to_prev(k + 1, BOSS::kSentinelCode);
+                if (prev_kmer != kmer) {
+                    // $**XF
+                    assert(kmer[0] == F);
+                    dummy_next_chunks[kmer[k]].add(kmer.data());
+                    prev_kmer = std::move(kmer);
+                }
+            };
+            std::vector<std::string> F_chunk_names(names.begin() + F * alphabet_size,
+                                                   names.begin() + (F + 1) * alphabet_size);
+            common::merge_files(F_chunk_names, write_dummy, true);
+
+            dummy_chunk.finish();
+            std::for_each(dummy_next_chunks.begin(), dummy_next_chunks.end(),
+                          [](auto &v) { v.finish(); });
+
+            #pragma omp atomic
+            num_kmers += dummy_chunk.size();
```

Comment:
Does declaring `num_kmers` to be `std::atomic<uint64_t>` make this slower?

Reply:
I didn't try, but I'm sure either way is fine here